ledger/src/blockstore_meta.rs

```diff
@@ -269,6 +304,339 @@ impl ShredIndex {
+        // - `bytes.len()` is checked to ensure it does not exceed `total_size`.
+        //
+        // 2. Pointer Safety:
+        // - `buffer` is a fresh `MaybeUninit`, ensuring no aliasing or invalid memory access.
+        // - `ptr.add(i)` is always within bounds as `i` comes from `bytes.chunks_exact(8).enumerate()`.
+        //
+        // 3. Alignment:
+        // - `buffer.as_mut_ptr()` is aligned for `u64`, matching the type of `ShredIndexV2Inner`.
+        // - Writing `u64` values respects alignment requirements.
+        //
+        // 4. Initialization:
+        // - Each `u64` in the buffer is initialized from 8-byte chunks of `bytes` using `u64::from_le_bytes`.
+        // - Remaining bytes are zeroed with `std::ptr::write_bytes`.
+        // - After the loop, the buffer is fully initialized, making `assume_init()` safe.
+        //
+        // 5. Compatibility:
+        // - Allows smaller inputs (`bytes.len() < total_size`) by zero-filling unused space to maintain backward compatibility.
+        unsafe {
+            let ptr = buffer.as_mut_ptr() as *mut u64;
+            // Write input bytes, converting to LE.
+            for (i, byte) in bytes.chunks_exact(8).enumerate() {
+                std::ptr::write(
+                    ptr.add(i),
+                    // SAFETY: chunks_exact(8) guarantees 8 bytes.
+                    u64::from_le_bytes(byte.try_into().unwrap_unchecked()),
+                );
+            }
+
+            // Zero remaining bytes
+            std::ptr::write_bytes(
+                (buffer.as_mut_ptr() as *mut u8).add(bytes.len()),
+                0,
+                remaining_bytes,
+            );
+
+            Ok(Self {
+                index: buffer.assume_init(),
+                num_shreds,
+            })
+        }
+    }
+}
+
+impl ShredIndexV2 {
+    pub fn num_shreds(&self) -> usize {
+        self.num_shreds
+    }
+
+    fn index_and_mask(index: u64) -> (usize, u64) {
+        // index / 64
+        let word_idx = (index >> 6) as usize;
+        // index % 64
+        let bit_idx = index & 63;
+        let mask = 1 << bit_idx;
+        (word_idx, mask)
+    }
+
+    #[cfg(test)]
+    fn remove(&mut self, index: u64) {
+        assert!(
+            index < MAX_DATA_SHREDS_PER_SLOT as u64,
+            "index out of bounds. {index} >= {MAX_DATA_SHREDS_PER_SLOT}"
+        );
+
+        let (word_idx, mask) = Self::index_and_mask(index);
+
+        if self.index[word_idx] & mask != 0 {
+            self.index[word_idx] ^= mask;
+            self.num_shreds -= 1;
+        }
+    }
+
+    #[allow(unused)]
+    pub(crate) fn contains(&self, idx: u64) -> bool {
+        if idx >= MAX_DATA_SHREDS_PER_SLOT as u64 {
+            return false;
+        }
+        let (word_idx, mask) = Self::index_and_mask(idx);
+        (self.index[word_idx] & mask) != 0
+    }
+
+    pub(crate) fn insert(&mut self, idx: u64) {
+        if idx >= MAX_DATA_SHREDS_PER_SLOT as u64 {
+            return;
+        }
```

Comment:
So just silently ignore if `idx >= MAX_DATA_SHREDS_PER_SLOT`?

This is one more place which can potentially break things if we change `MAX_DATA_SHREDS_PER_SLOT`.